src/charm.py

```
@@ -0,0 +1,595 @@
+        except Exception as err:  # pylint: disable=broad-exception-caught
+            logger.exception("Failed to reconcile runners")
+            self.unit.status = MaintenanceStatus(f"Failed to reconcile runners: {err}")
+            return
+
+        self.unit.status = ActiveStatus()
+
+    @catch_unexpected_action_errors
+    def _on_check_runners_action(self, event: ActionEvent) -> None:
+        """Handle the action of checking of runner state.
+
+        Args:
+            event: Action event of checking runner states.
+        """
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
+            event.fail("Missing token or org/repo path config")
+            return
+        if runner_manager.runner_bin_path is None:
+            event.fail("Missing runner binary")
+            return
+
+        online = 0
+        offline = 0
+        unknown = 0
+        runner_names = []
+
+        runner_info = runner_manager.get_github_info()
+
+        for runner in runner_info:
+            if runner.status == GitHubRunnerStatus.ONLINE.value:
+                online += 1
+                runner_names.append(runner.name)
+            elif runner.status == GitHubRunnerStatus.OFFLINE.value:
+                offline += 1
+            else:
+                # might happen if runner dies and GH doesn't notice immediately
+                unknown += 1
+        event.set_results(
+            {
+                "online": online,
+                "offline": offline,
+                "unknown": unknown,
+                "runners": ", ".join(runner_names),
+            }
+        )
+
+    @catch_unexpected_action_errors
+    def _on_reconcile_runners_action(self, event: ActionEvent) -> None:
+        """Handle the action of reconcile of runner state.
+
+        Args:
+            event: Action event of reconciling the runner.
+        """
+        runner_manager = self._get_runner_manager()
+        if not runner_manager:
```

Perhaps get runner manager can raise an exception instead of returning `None` which provides the error message? It seems this message is repeated a lot

---

Perhaps injecting the runner manager can even be a decorator

---

Good idea

---

Minor comments